packages/near-lake-primitives/src/types/block.ts

```diff
@@ -0,0 +1,159 @@
+        }
+        return this.executedReceipts;
+    }
+
+    actions(): Action[] {
+        const actions: Action[] = this.streamerMessage.shards.flatMap((shard) => shard.receiptExecutionOutcomes).filter((exeuctionOutcomeWithReceipt) => isActionReceipt(exeuctionOutcomeWithReceipt.receipt)).map((exeuctionOutcomeWithReceipt) => receiptViewToAction(exeuctionOutcomeWithReceipt.receipt)).filter((action) => action !== null).map(action => action as Action)
+        return actions
+    }
+
+    events(): Event[] {
+        const events = this.receipts().flatMap((executedReceipt) => executedReceipt.logs.map(logToRawEvent).map((rawEvent) => {
+            if (rawEvent) {
+                let event = { relatedReceiptId: executedReceipt.receiptId, rawEvent: rawEvent } as Event
+                return event
+            } else {
+                return null
+            }
+        })).filter((event) => event !== null).map((event) => event as Event)
+        return events
+    }
+    stateChanges(): StateChange[] {
+        if (this._stateChanges.length == 0) {
+            this._stateChanges = this.streamerMessage.shards.flatMap((shard) => shard.stateChanges).map(fromStateChangeViewToStateChange)
+        }
+        return this._stateChanges
+    }
+    actionByReceiptId(receipt_id: string): Action | undefined {
+        if (this._actions.size == 0) {
+            this.buildActionsHashmap()
+        }
+        return this._actions.get(receipt_id);
+    }
+
+    eventsByReceiptId(receipt_id: string): Array<Event> {
+        if (this._events.size == 0) {
+            this.buildEventsHashmap()
+        }
+        return this._events.get(receipt_id) || [];
+    }
+
+    eventsByAccountId(account_id: string): Array<Event> {
+        return this.events().filter((event) => {
+            const action = this.actionByReceiptId(event.relatedReceiptId)
+            if (action !== undefined && action?.receiverId == account_id || action?.signerId == account_id) {
+                return true
+            }
+            return false
+        });
+    }
+
+    private buildActionsHashmap() {
+        const actions = new Map<string, Action>();
+        this.actions().forEach(action => {
+            actions.set(action.receiptId, action)
+        }
+        );
+        this._actions = actions
+    }
+
+    private buildEventsHashmap(): Map<string, Array<Event>> {
+        const events = new Map<string, Array<Event>>();
+        for (const receipt of this.executedReceipts) {
+            events.set(receipt.receiptId, receipt.events);
+        }
+        return events;
+    }
+
+
+}
+
+export function createBlock(streamerMessage: StreamerMessage): Block {
+    const block: Block = new Block(streamerMessage, [], [], [], new Map(), new Map(), []);
+    return block;
+}
+
+export type BlockHeader = {
+    height: number;
```

Comment:
Sorry if my comment was too early and you haven't worked on that yet.

You've decided to replicate the getters approach as I did in Rust Lake Primitives. Getters-approach is used on each structure in Rust Lake Primitives, so I expect to see the same here.

Reply:
In my first approach, I actually had most of these structures represented as classes with getters in place. I noticed that for a lot of these structures, we simply return the fields and do not perform any transformations on those fields, so I opted for a simpler approach of making these simple structures into types instead. 

One thing frustrating about turning these constructors into a class is the problem with naming and autocomplete suggestions for types in Typescript that I noticed. @morgsmccauley, youy might be able to shed some light here. 

To give you an example, this is what BlockHeader looks like as a class with getters in Typescript(it is only half of the structure)

```Typecript
export class BlockHeader {
    private _height: number;
    private _hash: string;
    private _prevHash: string;
    private _author: string;
    private _timestampNanosec: string;
    private _epochId: string;
    constructor(height: number, hash: string, prevHash: string, author: string, timestampNanosec: string, epochId: string) {
        this._height = height;
        this._hash = hash;
        this._prevHash = prevHash;
        this._author = author;
        this._timestampNanosec = timestampNanosec;
        this._epochId = epochId;
    }
    get height() {
        return this._height
    }
    get hash() {
        return this._hash
    }
    get prevHash() {
        return this._prevHash
    }
    get author() {
        return this._author
    }
    get timestampNanosec() {
        return this._timestampNanosec
    }
    get epochId() {
        return this._epochId
    }
}

/// This is how we would create a BlockHeader.
new BlockHeader(1, "hash", "prevHash", "author", "timestampNanosec", "epochId")

```
Notice that we are not able to pass in named parameters into the Blockheader to create this object which I thought was a pretty bad trade off compared to using an interface. I'm sure that there may be hacky way to be able to accomplish named parameters for instantiating objects though(more on this later). It just makes thing more clunky. 

As supposed to keeping the structure as a Type, we can do the following and define the object with their named types. 
```
    const blockHeader: BlockHeader = {
        height: streamerMessage.block.header.height,
        hash: streamerMessage.block.header.hash,
        prevHash: streamerMessage.block.header.prevHash,
        author: streamerMessage.block.author,
        timestampNanosec: streamerMessage.block.header.timestampNanosec,
        epochId: streamerMessage.block.header.epochId,
    };
```

But we lose the ability to create helper functions on this structure. 

**Another approach**

Define a type and a class that goes along with that type. 

```
export type TBlockHeader = {
    height: number;
    hash: string;
    prevHash: string;
    author: string;
    timestampNanosec: string;
    epochId: string;
}
class BlockHeader {
    private _height: number;
    private _hash: string;
    private _prevHash: string;
    private _author: string;
    private _timestampNanosec: string;
    private _epochId: string;
    constructor(header: TBlockHeader) {
        this._height = header.height;
        this._hash = header.hash;
        this._prevHash = header.prevHash;
        this._author = header.author;
        this._timestampNanosec = header.timestampNanosec;
        this._epochId = header.epochId;
    }
    get height() {
        return this._height
    }
    get hash() {
        return this._hash
    }
    get prevHash() {
        return this._prevHash
    }
    get author() {
        return this._author
    }
    get timestampNanosec() {
        return this._timestampNanosec
    }
    get epochId() {
        return this._epochId
    }

    // some helper function we decide to put in later. 
    helperFunction() {
        // do something
    }
}

```

This way, we can instantiate the object using named parameters, 

```
new BlockHeader({ height: 1, hash: "hash", prevHash: "prevHash", author: "author", timestampNanosec: "timestampNanosec", epochId: "epochId" }).helperFunction

```

as well as able to create helper methods inside the class if needed later. 

Still, I do not think every structure should be a class. We should leave structures that we think we won't modify as types and only keep classes for structures that we think we may need to make more convenient by adding helper methods to them.

Reply:
Agree with @roshaans, I only think `class` is necessary if:
1. More logic is required within the `get` methods, i.e. they don't just return the data in the same form it was passed in
2. There are methods which operate on the data which should therefore be defined with the data itself

Looking at the code a bit closer, I'm starting to think turning `BlockHeader` in to a class wouldn't be a bad idea. `streamerMessageToBlockHeader()` could then be defined as a `static` method on the class to allow for easy initialisation, and keeping data/functionality close like I've mentioned above in 2. This also allows for easy adding of additional functionality in future, which can be defined on the class.

We wouldn't need to define `get`ters, just make them all `public`/`readonly`. 

> One thing frustrating about turning these structures into a class is the problem with naming and autocomplete suggestions for types in Typescript that I noticed

@roshaans I understand the frustration, but IMO I would prefer separate parameters for the following reasons:
1. Most IDE's should tell you the types of each argument required in a class constructor, or allow you to jump to the definition to inspect what is necessary. There shouldn't isn't really any overhead compared to named parameters.
2. There are a number of TS features geared towards separate parameters in class constructors, e.g. Parameter Properties like I've suggested before.

Reply:
I totally rely on you guys here. I want to remind the choice of getters approach in Rust. 

The intention was:
As of now, we simply return the value of the fields. However, I keep in mind that we might need to change some logic or even internal fields (names or even their number). So I want to avoid breaking the API for the developers as much as possible. IMO it might look overcomplicated in the beginning but may help us in the future.

Reply:
Thank you for the review @morgsmccauley! I converted most if not all the conversion methods from one type to another into a Static method on that target class type. That makes things much cleaner. 

@khorolets At the moment, A class exists for `Block`, `BlockHeader`, `Receipt`, `Action`, and `StateChange`.  When and if we decide to add getters to these classes, this should not change the API for the devs. So I think this should be fine.

Reply:
I decided to make the majority of structures into Classes after learning about a shorthand way to create classes from @morgsmccauley. Thanks!